src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+                        offset += 4;
+                        optionDataSize = 4;
+                        break;
+
+                    case (int)PreLoginOptions.MARS:
+                        payload[payloadLength++] = (byte)(context.ConnectionContext.ConnectionString.MARS ? 1 : 0);
+                        offset += 1;
+                        optionDataSize += 1;
+                        break;
+
+                    case (int)PreLoginOptions.TRACEID:
+                        context.ConnectionContext.ConnectionId.TryWriteBytes(payload.AsSpan(payloadLength, GUID_SIZE));
+                        payloadLength += GUID_SIZE;
+                        offset += GUID_SIZE;
+                        optionDataSize = GUID_SIZE;
+
+                        ActivityCorrelator.ActivityId actId = ActivityCorrelator.Next();
+                        actId.Id.TryWriteBytes(payload.AsSpan(payloadLength, GUID_SIZE));
+                        payloadLength += GUID_SIZE;
+                        payload[payloadLength++] = (byte)(0x000000ff & actId.Sequence);
+                        payload[payloadLength++] = (byte)((0x0000ff00 & actId.Sequence) >> 8);
+                        payload[payloadLength++] = (byte)((0x00ff0000 & actId.Sequence) >> 16);
+                        payload[payloadLength++] = (byte)((0xff000000 & actId.Sequence) >> 24);
+                        int actIdSize = GUID_SIZE + sizeof(uint);
+                        offset += actIdSize;
+                        optionDataSize += actIdSize;
+                        SqlClientEventSource.Log.TryTraceEvent("<sc.TdsParser.SendPreLoginHandshake|INFO> ClientConnectionID {0}, ActivityID {1}",
+                            context.ConnectionContext.ConnectionId, actId);
+                        break;
+
+                    case (int)PreLoginOptions.FEDAUTHREQUIRED:
+                        payload[payloadLength++] = 0x01;
+                        offset += 1;
+                        optionDataSize += 1;
+                        break;
+
+                    default:
+                        Debug.Fail("UNKNOWN option in SendPreLoginHandshake");
+                        break;
+                }
+
+                await tdsStream.WriteByteAsync((byte)((optionDataSize & 0xff00) >> 8), isAsync, ct).ConfigureAwait(false);
+                await tdsStream.WriteByteAsync((byte)(optionDataSize & 0x00ff), isAsync, ct).ConfigureAwait(false);
+            }
+
+            // Write out last option - to let server know the second part of packet completed
+            await tdsStream.WriteByteAsync((byte)PreLoginOptions.LASTOPT, isAsync, ct).ConfigureAwait(false);
+
+            if (isAsync)
+            {
+                // Write out payload
+                await tdsStream.WriteAsync(payload.AsMemory(0, payloadLength), ct).ConfigureAwait(false);
+            }
+            else
+            {
+                tdsStream.Write(payload.AsSpan(0, payloadLength));
```

Comment:
The `CancellationToken` needs to throw if cancelled here. The same also applies for the synchronous `Flush` too.

Reply:
Or call WriteAsync() with isAsync + cancellationToken

Reply:
There is no WriteAsync for Byte[] overload with isAsync/ct The WriteAsync mimics the overload for Streams.